File: files_and_users/java/src/main/java/com/nublic/filesAndUsers/java/Mirror.java
package com.nublic.filesAndUsers.java;

import java.util.ArrayList;
import java.util.List;

import org.freedesktop.dbus.exceptions.DBusException;

public class Mirror extends Folder {
	static String DATA_ROOT = "/var/nublic/data/mirrors/";
	int id;
	
	public Mirror(int id) {
		this.id = id;
	}
	
	public static List<Mirror> getAll() throws FileQueryException {
		try {
			ArrayList<Mirror> mirrors = new ArrayList<Mirror>();
			String m = Singletons.getMirrors().get_all_mirrors();
			if (m.isEmpty())
				return mirrors; 
			String[] ids = m.split(":");
			for (String mid : ids) {
				mirrors.add(new Mirror(Integer.parseInt(mid)));
			}
			return mirrors;
		} catch(DBusException e) {
			throw new FileQueryException();
		}
	}
	
	public boolean exists() throws FileQueryException {
		try {
			return Singletons.getMirrors().mirror_exists(id);
		} catch(DBusException e) {
			throw new FileQueryException();
		}
	}
	
	public int getId() {
		return this.id;
	}
	
	public String getName() throws FileQueryException {
		try {
			return Singletons.getMirrors().get_mirror_name(id);
		} catch(DBusException e) {
			throw new FileQueryException();
		}
	}
	
	public void setName(String name) throws FileQueryException {
		try {
			Singletons.getMirrors().change_mirror_name(id, name);
		} catch(DBusException e) {
			throw new FileQueryException();
		}
	}
	
	public User getOwner() throws FileQueryException {
		try {
			return new User(Singletons.getMirrors().get_mirror_owner(id));
		} catch(DBusException e) {
			throw new FileQueryException();
		}
	}
	
	public static Mirror create(String name, String owner) throws FileQueryException {
		try {
			return new Mirror(Singletons.getMirrors().create_mirror(name, owner));
		} catch(DBusException e) {
			throw new FileQueryException();
		}
	}
	
	public void delete(boolean removeInFileSystem) throws FileQueryException {
		try {
			Singletons.getMirrors().delete_mirror(id, removeInFileSystem);
		} catch(DBusException e) {
			throw new FileQueryException();
		}
	}
	
	public String getPath() {
		return DATA_ROOT + id;
	}
}

File: python/client/src/nublic/files_and_users/mirror.py

from ..dbus_in_other_thread import call_expecting_return, call_without_return
from ..rpcbd_client import rpcbd_call_return, rpcbd_call
import string

import user

MIRROR_DATA_ROOT = "/var/nublic/data/mirrors/"

def _call_mirror_method_return(m, use_dbus=False):
    if use_dbus:
        return call_expecting_return('com.nublic.files', '/com/nublic/Mirrors', 'com.nublic.files', m)
    else:
        return rpcbd_call_return(5440, m)

def _call_mirror_method(m, use_dbus=False):
    if use_dbus:
        return call_without_return('com.nublic.files', '/com/nublic/Mirrors', 'com.nublic.files', m)
    else:
        return rpcbd_call(5440, m)

def get_all_mirrors(use_dbus=False):
    notSplitted = _call_mirror_method_return(lambda i: i.get_all_mirrors(), use_dbus)
    if notSplitted == '':
        return []
    # Take elements apart
    splitted = string.split(notSplitted, ':')
    mirrorList = []
    for midS in splitted:
        mirrorList.append(Mirror(int(midS), use_dbus))
    return mirrorList

def create_mirror(name, owner, use_dbus=False):
    mid = _call_mirror_method_return(lambda i: i.create_mirror(name, owner), use_dbus)
    return Mirror(mid)

class Mirror:
    '''
    Represents a mirror in the Nublic file system
    '''
    def __init__(self, mid, use_dbus=False):
        self._id = mid
        self.use_dbus = use_dbus
    
    def exists(self):
        return _call_mirror_method_return(lambda i: i.mirror_exists(self._id), self.use_dbus)
    
    def get_id(self):
        return self._id
    
    def get_name(self):
        return _call_mirror_method_return(lambda i: i.get_mirror_name(self._id), self.use_dbus)
    
    def change_name(self, name):
        return _call_mirror_method_return(lambda i: i.change_mirror_name(self._id, name), self.use_dbus)
    
    def get_owner(self):
        return user.User(_call_mirror_method_return(lambda i: i.get_mirror_owner(self._id), self.use_dbus), self.use_dbus)
    
    def get_path(self):
        return MIRROR_DATA_ROOT + str(self._id)
    
    def delete(self, remove_in_file_system):
        _call_mirror_method(lambda i: i.delete_mirror(self._id, remove_in_file_system), self.use_dbus)
    
    def can_be_read_by(self, usr):
        return usr.can_read(self.get_path())
    
    def can_be_written_by(self, usr):
        return usr.can_write(self.get_path())
    
